pkg/agent/secondarynetwork/podwatch/controller.go

```diff
@@ -0,0 +1,402 @@
+	netdefutils "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/utils"
+
+	cniserver "antrea.io/antrea/pkg/agent/cniserver"
+	cnipodcache "antrea.io/antrea/pkg/agent/secondarynetwork/cnipodcache"
+	ipam "antrea.io/antrea/pkg/agent/secondarynetwork/ipam"
+)
+
+const (
+	controllerName = "AntreaAgentPodController"
+	minRetryDelay  = 2 * time.Second
+	maxRetryDelay  = 120 * time.Second
+	numWorkers     = 4
+)
+
+const (
+	networkAttachDefAnnotationKey = "k8s.v1.cni.cncf.io/networks"
+	cniPath                       = "/opt/cni/bin/"
+	sriovInterfaceType            = "sriov"
+	defaultSecondaryInterfaceName = "eth1"
+	podDelComplete                = 0
+)
+
+// Set resyncPeriod to 0 to disable resyncing.
+const resyncPeriod = 0 * time.Minute
+
+type PodController struct {
+	kubeClient         clientset.Interface
+	netAttachDefClient netdefclient.K8sCniCncfIoV1Interface
+	queue              workqueue.RateLimitingInterface
+	podInformer        cache.SharedIndexInformer
+	podLister          corelisters.PodLister
+	nodeName           string
+	podCache           cnipodcache.CNIPodInfoStore
+	cniServer          *cniserver.CNIServer
+}
+
+func NewPodController(kubeClient clientset.Interface,
+	netAttachDefClient netdefclient.K8sCniCncfIoV1Interface,
+	podInformer cache.SharedIndexInformer,
+	resyncPeriod time.Duration,
+	nodeName string,
+	podCache cnipodcache.CNIPodInfoStore,
+	cniServer *cniserver.CNIServer) *PodController {
+
+	pc := PodController{
+		kubeClient:         kubeClient,
+		netAttachDefClient: netAttachDefClient,
+		queue:              workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "podcontroller"),
+		podInformer:        podInformer,
+		podLister:          corelisters.NewPodLister(podInformer.GetIndexer()),
+		nodeName:           nodeName,
+		podCache:           podCache,
+		cniServer:          cniServer,
+	}
+
+	podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    pc.enqueuePod,
+			DeleteFunc: pc.enqueuePod,
+			UpdateFunc: func(old, cur interface{}) { pc.enqueuePod(cur) },
+		},
+		resyncPeriod,
+	)
+	return &pc
+}
+
+func podKeyGet(pod *corev1.Pod) string {
+	return pod.Namespace + "/" + pod.Name
+}
+
+func (pc *PodController) enqueuePod(obj interface{}) {
+	var err error
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		podDeletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(err, "Unexpected object received:", obj)
+			return
+		}
+		pod, ok := podDeletedState.Obj.(*corev1.Pod)
+		if !ok {
+			klog.ErrorS(err, "DeletedFinalStateUnknown object is not of type Pod: ", podDeletedState.Obj, pod)
+			return
+		}
+	}
+	podKey := podKeyGet(pod)
+	pc.queue.Add(podKey)
+}
+
+// handleAddUpdatePod handles Pod Add, Update events and updates annotation if required.
+func (pc *PodController) handleAddUpdatePod(obj interface{}) error {
```

Comment:
I asked before. Could you explain to me what will happen if we handle Pod update event twice. Will we recreate an interface for every secondary network in the Pod annotation? The same question applies to partial failure of secondary interface creation - if we retry what will happen for interfaces had been created earlier?

I proposed if this is an issue, can we add a boolean flag in podCNIInfo - like secondaryNetworkCreated - which indicates at least one secondary network has been created for this Pod, and then we check here if it is true we ignore the further Pod update events, as we do not support updates today.